Specific messages do not show up in retrieved message history

I've observed that with the plugin v1.8.0, specific messages that I know were added to a MUC room didn't show up in the message history as retrieved through MAM2. These messages didn't show up in both Converse (4.2.0) and Conversations.

One of the messages that was missing, was this one:

> Something like this: Spark.exe username=bart&password=test&domain=push04.goodbytes.im

(That domain nor that account exists any longer, so there's no point in trying to abuse those credentials :-) )

I'm guessing that the monitoring plugin tries to parse the `&password` text as some kind of entity reference, as I've noticed that the following was logged to the log of Openfire. Openfire should either escape text like that, or not parse entities.

```2019.07.03 21:13:58 ERROR [message-archive-handler-1035]: com.reucon.openfire.plugin.archive.xep0313.IQQueryHandler - Failed to parse message stanza.
org.dom4j.DocumentException: Error on line 3 of document  : The reference to entity "password" must end with the ';' delimiter.
        at org.dom4j.io.SAXReader.read(SAXReader.java:462) ~[dom4j-2.1.1.jar:?]
        at org.dom4j.DocumentHelper.parseText(DocumentHelper.java:286) ~[dom4j-2.1.1.jar:?]
        at com.reucon.openfire.plugin.archive.xep0313.IQQueryHandler.sendMessageResult(IQQueryHandler.java:376) ~[?:?]
        at com.reucon.openfire.plugin.archive.xep0313.IQQueryHandler.lambda$handleIQ$0(IQQueryHandler.java:207) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_201]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_201]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_201]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_201]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_201]
Caused by: org.xml.sax.SAXParseException: The reference to entity "password" must end with the ';' delimiter.
        at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:203) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:177) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:400) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1472) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEntityReference(XMLDocumentFragmentScannerImpl.java:1850) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:3061) ~[?:1.8.0
_201]
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:602) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:505) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:842) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:771) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1213) ~[?:1.8.0_201]
        at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:643) ~[?:1.8.0_201]
        at org.dom4j.io.SAXReader.read(SAXReader.java:445) ~[dom4j-2.1.1.jar:?]
        ... 8 more
```

--- a/changelog.html
+++ b/changelog.html
@@ -44,6 +44,11 @@
 Monitoring Plugin Changelog
 </h1>
 
+<p><b>1.8.2</b> -- (to be determined)</p>
+<ul>
+    <li>[<a href='https://github.com/igniterealtime/openfire-monitoring-plugin/issues/46'>Issue #46</a>] - Specific messages do not show up in retrieved message history.</li>
+</ul>
+
 <p><b>1.8.1</b> -- September 25, 2019</p>
 <ul>
     <li>[<a href='https://github.com/igniterealtime/openfire-monitoring-plugin/issues/53'>Issue #53</a>] - XSS in Monitoring Plugin's Archiving Settings page.</li>

--- a/src/java/com/reucon/openfire/plugin/archive/impl/MucMamPersistenceManager.java
+++ b/src/java/com/reucon/openfire/plugin/archive/impl/MucMamPersistenceManager.java
@@ -6,9 +6,7 @@ import com.reucon.openfire.plugin.archive.model.ArchivedMessage;
 import com.reucon.openfire.plugin.archive.model.Conversation;
 import com.reucon.openfire.plugin.archive.model.Participant;
 import com.reucon.openfire.plugin.archive.xep0059.XmppResultSet;
-import org.dom4j.Document;
-import org.dom4j.DocumentException;
-import org.dom4j.DocumentHelper;
+import org.dom4j.*;
 import org.jivesoftware.database.DbConnectionManager;
 import org.jivesoftware.openfire.XMPPServer;
 import org.jivesoftware.openfire.muc.MUCRoom;
@@ -34,6 +32,7 @@ import java.util.*;
  */
 public class MucMamPersistenceManager implements PersistenceManager {
     private final static Logger Log = LoggerFactory.getLogger( MucMamPersistenceManager.class );
+    protected static final DocumentFactory docFactory = DocumentFactory.getInstance();
     private static final int DEFAULT_MAX = 100;
 
     @Override
@@ -179,7 +178,10 @@ public class MucMamPersistenceManager implements PersistenceManager {
 
                         if ( pos1 > -1 && pos2 > -1 )
                         {
-                            stanza = stanza.substring( 0, pos1 + 6 ) + body + stanza.substring( pos2 );
+                            // Add the body value to a proper XML element, so that the strings get XML encoded (eg: ampersand is escaped).
+                            final Element bodyEl = docFactory.createDocument().addElement("body");
+                            bodyEl.setText(body);
+                            stanza = stanza.substring( 0, pos1 ) + bodyEl.asXML() + stanza.substring( pos2 + 7 );
                         }
                     }
                     final Document doc = DocumentHelper.parseText( stanza );

